docs/shapes.md

```diff
@@ -0,0 +1,193 @@
+
+```js
+// What the getter/setter are not important!
+Object.defineProperty(obj, "y", {
+  enumerable: true,
+  configurable: true,
+  get: function () {
+    return this.x;
+  },
+  set: function (value) {
+    this.x = value;
+  },
+});
+```
+
+```mermaid
+flowchart TD
+    Root(S0: Root Shape\n\nPrototype: None) --->| Property Count: 0 | PropertyTable[PropertyTable\n\nx: Slot 0, writable, configurable, enumerable\ny: Slot 1, has_get, has_set, configurable, enumerable]
+
+    ObjectPrototype(S1: Prototype Shape\n\nPrototype: Object.prototype) -->| Property Count: 0 |PropertyTable
+    ObjectPrototype -->|Previous| Root
+
+    InsertX(S2: Insert Shape\n\nPrototype: Object.prototype) --> | Property Count: 1 | PropertyTable
+    InsertX -->|Previous| ObjectPrototype
+
+    InsertY(S3: Insert Shape\n\nPrototype: Object.prototype) --> | Property Count: 2 | PropertyTable
+    InsertY -->|Previous| InsertX
+
+    O(Object o\n\nElement 0: JsValue: 100\nElement 1: JsValue: func\nElement 2: JsValue: func) --> InsertY
+```
+
+We can see that the property has been added into the property table, it has the `has_get` and `has_set` flags set,
+in the object there are two elements added, the first is the `get` function and the second is the `set` function.
+
+Slots are varying in length, two for accessor properties and one for data properties, the index points to the first
+value in the object storage.
+
+What would happen if an object had `S2` shape and we tried to access a property `y` how does it know if it
+has or doesn't have a property named `y`? By the property count on the shape, it has property count `1`,
+all the object in the `PropertyTable` are stored in a map that preserves the order and and can be indexed.
+
+When we do a lookup the on property table, if the index of the property is greater than the property count (`1`),
+than it does not belong to the shape.
+
+Now, Let's create a new object `o2`, with property `x`:
+
+```js
+let o2 = { x: 200 };
+```
+
+After this `o2` would have the `S2` shape.
+
+How does the shape know that it can reuse `S1` then to go to `S2`? This is not the real structure!
+Every shape has pointers to forward transitions that happened, these are weak pointers so we don't keep
+alive unused shapes. The pointers have been omitted, so the diagrams are clearer (too many arrows).
+
+Ok, now let us define a property `z` instead of `y`:
+
+```js
+o2.z = 300;
+```
+
+The following changes accure to the shape tree:
+
+```mermaid
+flowchart TD
+    Root(S0: Root Shape\n\nPrototype: None) --->| Property Count: 0 | PropertyTable[PropertyTable\n\nx: Slot 0, writable, configurable, enumerable\ny: Slot 1, has_get, has_set, configurable, enumerable]
+
+    ObjectPrototype(S1: Prototype Shape\n\nPrototype: Object.prototype) -->| Property Count: 0 |PropertyTable
+    ObjectPrototype -->|Previous| Root
+
+    InsertX(S2: Insert Shape\n\nPrototype: Object.prototype) --> | Property Count: 1 | PropertyTable
```

Comment:
<img width="470" alt="image" src="https://user-images.githubusercontent.com/936006/228367294-5f6dc50e-79db-4f48-8d6f-03843a2a0b96.png">

It may be worth showing which property was added, as currently Shape 2 and Shape 3 look the same, apart from the "property add" on the line. It doesn't seem obvious what was added.

Maybe you could put in the box ("added x", "added y") or something